src/Access/KerberosInit.cpp

```diff
@@ -0,0 +1,213 @@
+#include <Access/KerberosInit.h>
+#include <Common/Exception.h>
+#include <Common/logger_useful.h>
+#include <Poco/Logger.h>
+#include <Loggers/Loggers.h>
+#include <filesystem>
+#if USE_KRB5
+#include <krb5.h>
+#include <mutex>
+
+using namespace DB;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int KERBEROS_ERROR;
+}
+}
+
+struct k5_data
+{
+    krb5_context ctx;
+    krb5_ccache in_cc, out_cc;
+    krb5_principal me;
+    char * name;
+    krb5_boolean switch_to_cache;
+};
+
+class KerberosInit
+{
+public:
+    int init(const String & keytab_file, const String & principal, const String & cache_name = "");
+    ~KerberosInit();
+private:
+    struct k5_data k5;
+    krb5_ccache defcache = nullptr;
+    krb5_get_init_creds_opt * options = nullptr;
+    krb5_creds my_creds;
+    krb5_keytab keytab = nullptr;
+    krb5_principal defcache_princ = nullptr;
+};
+
+int KerberosInit::init(const String & keytab_file, const String & principal, const String & cache_name)
+{
+    auto log = &Poco::Logger::get("KerberosInit");
+    LOG_DEBUG(log,"Trying to authenticate to Kerberos v5");
+
+    krb5_error_code ret;
+
+    const char *deftype = nullptr;
+    int flags = 0;
+
+    if (!std::filesystem::exists(keytab_file))
+        throw Exception("Error keytab file does not exist", ErrorCodes::KERBEROS_ERROR);
```

Comment:
I presume that word "Error" woud be duplicated, it is redundant in error message.
What is visible for user?

Reply:
Corrected error message by removing Error text.